File: practice/20_Valid_Parentheses.kt
fun main() {
	println(valid("()[]{}"))
	println(valid("([])"))
	println(valid("([]"))
}

fun valid(s: String): Boolean {
	var keys = mapOf(')' to '(', ']' to '[', '}' to '{')
	var stack = mutableListOf<Char>()

	for(char in s) {
		if(stack.isNotEmpty() && keys[char] == stack.last()) {
			stack.removeAt(stack.size-1)
			continue
		}
		stack.add(char)
	}

	return stack.isEmpty()
}

File: practice/20_Valid_Parentheses.js
const keys = {
    ')': '(',
    ']': '[',
    '}': '{',
};

function valid(s) {
    let stack = [];

    for (let i = 0; i < s.length; i++) {
        if (stack.length > 0 && keys[s[i]] === stack[stack.length - 1]) {
            stack.pop();
            continue;
        }
        stack.push(s[i]);
    }

    return !stack.length > 0;
}

console.log(valid('()[]{}'));
console.log(valid('([])'));
console.log(valid('([]'));
